### DISCUSSION TITLE
Why  Span<MyStruct> arr = stackalloc MyStruct[10]; allocated memory on heap?

### DISCUSSION THREAD

**Discussion by P9avel** [dotnet/runtime#93895] [General] [answered]
Hi. I have code
public struct MyStruct
{
    public MyStruct()
    {
        x = 1;
        y = 2;
        z = 3;
    }

    public int x;
    public int y;
    public int z;
}

internal class Program
{
    static void Main(string[] args)
    {

        Span<MyStruct> arr = stackalloc MyStruct[10];
        Console.ReadKey();
    }
}

I am dont understood why stackalloc MyStruct[10] allocated memory on heap. Can you to describe?

**Comment by svick** [comment]
What makes you think it allocates memory on the heap?

**Comment by P9avel** [comment]
Please look screenshot
![image](https://github.com/dotnet/runtime/assets/95086762/a3356e02-bfa2-4ef3-aab1-348bf7c3f5e5)
I am dont undestand why i see MyStruct instances in heap viewer

**Reply by svick** [reply]
That looks like an artifact of the debugger. Notice that you also see `SpanDebugView<MyStruct>` there, which is used by the debugger to visualize spans and [allocates arrays to do that](https://github.com/dotnet/runtime/blob/a7282289/src/libraries/System.Private.CoreLib/src/System/SpanDebugView.cs).

**Comment by P9avel** [comment]
Thanks you for your comment. Possible exists visual tool/pluggin for debugging stack?

**Reply by Clockwork-Muse** [reply, marked as answer]
What is it, exactly, that you're trying to do?  You can already see the contents of things on the stack as part of normal debugging, even if it isn't always made clear that they're there (eg, local variables).
Most things on the stack are going to be there are there by definition of what they are (structs, stackalloc-ed variables, etc), so something that additionally lists them "on stack" isn't terribly helpful (unlike heap, where items can live hide or persist in some situations).

**Reply by P9avel** [reply]
Thank you for your detail description. Very helped for me
